### ISSUE TITLE
Enumeration's type-checks in production builds

### ISSUE THREAD

**Issue by buffaybu** [mobxjs/mobx-state-tree#600]
In production builds, if an enumeration property is assigned with a value not in that enumeration, MST will raise an error.

For example, in mst-example-bookshop add these lines:

```js
export const Book = types.model("Book", {
    // ...,
    publisher: types.enumeration(["Penguin"])  // add this
})
// ...
        function updateBooks(json) {
            // ...
            json.forEach(bookJson => {
                bookJson.publisher = 'Turtle' // add this
                ...
            })
        }
```

In development env, MST will say: 

> ... at path "/publisher" value `"Turtle"` is not assignable to type: `("Penguin")` (No type is applicable for the union), expected an instance of `("Penguin")` or a snapshot like `("Penguin")` instead.

That is fine.

But in production:

> Failed to load books  TypeError: Cannot read property 'instantiate' of undefined

Maybe in production it's better to be able to assign any value to an enumeration property or a union property.

**Comment by mattiamanzati** [comment]
By default runtime typechecking is turned off in production mode, you can import the function typecheck function from the MST package and manually perform the typechecking over the json received from the server :)

**Comment by buffaybu** [comment]
Sorry. The original title may be misleading.

What I mean is that the typechecking for enumerations seems not turned off in production mode and there is an error raised.

Actually the error is from [union.ts#L50](https://github.com/mobxjs/mobx-state-tree/blob/85b5edd7bd9691ffde59b404b429285ee6a3fcf6/packages/mobx-state-tree/src/types/utility-types/union.ts#L50). It assumes that it will always find the type from the union, but in production it may not.

**Comment by mattiamanzati** [comment]
Yeah, that's expected because you are putting an illegal value, so it will error for sure :) What behaviour did you expected?

**Comment by buffaybu** [comment]
In my project I'm fetching items from an external API. An item has several status so the item store has a property `status` as an enumeration. But the API is not stable and in some cases `status` may be some other values.

My thought is that setting an unclaimed value/type to an enumeration/union is a type error, and since typechecking is turned off in production mode, it's ok to set that value in production.

Furthermore, there is no enum or union type in vanilla JS so it feels normal that these types should just take whatever they receive and infer nothing in production. Just leave the inference in dev mode.

Are you suggesting that a union is still a union in production, so in this case it's better to just use `types.string`? 😶

**Comment by mattiamanzati** [comment]
Yep, definetely the latter one if you expect the value to be set from whathever the rest api says :)

**Comment by buffaybu** [comment]
Um..🤔

That feels right. Thanks for your time.

**Comment by mweststrate** [comment]
Thanks for spotting! It is not technically incorrect, but easily confusing
indeed. Would you mind adding a null check there and `fail` if appropriate?

Op ma 15 jan. 2018 om 09:46 schreef BuffayBu <notifications@github.com>:
